### PR TITLE
Upgrade macOS/Travis default image to xcode6.4

### PR DESCRIPTION
**Pull request by inducer**
Here's a quick straw man for a CFEP to get the Travis image changed.

cc @jakirkham @conda-forge/core

### CODE REVIEW

**Comment by inducer** [comment]
[Pretty version](https://github.com/inducer/conda-forge-enhancement-proposals/blob/master/cfep-02.md)

File: cfep-02.md
Comment (mwcraig) [inline cfep-02.md]: @inducer -- here is the ref: https://github.com/travis-ci/travis-ci/issues/6765#issuecomment-256703076

**Comment by mwcraig** [comment]
This seems inevitable to me. One reasonable approach to testing (not saying I actually have time to implement it, unfortunately) might be to build a couple test packages on xcode 6.4 that target 10.9, and use the xcode 6.1 while it is still around to try running test suites for those packages. 

That won't catch everything, but at some point we are really constrained in what we can do by what the CI services make it easy to do.

**Comment by inducer** [comment]
Matt Craig notifications@github.com writes:

> @inducer -- here is the ref: https://github.com/travis-ci/travis-ci/issues/6765#issuecomment-256703076

Thanks, updated!

File: cfep-02.md
Comment (jakirkham) [inline cfep-02.md:82]: Other relevant references.
- [Travis CI changes default to Mac OS 10.11 XCode 7.3](https://blog.travis-ci.com/2016-10-04-osx-73-default-image-live/)
- [Travis CI image restructuring](https://blog.travis-ci.com/2016-09-15-new-default-osx-image-coming/)
- [Currently available Travis CI OS X images](https://docs.travis-ci.com/user/osx-ci-environment/#OS-X-Version)
- [Details on Mac OS 10.9 XCode 6.1](https://docs.travis-ci.com/user/osx-ci-environment/#Xcode-6.1) supports 10.9 and 10.10
- [Details on Mac OS 10.10 XCode 6.4](https://docs.travis-ci.com/user/osx-ci-environment/#Xcode-6.4) supports 10.9 and 10.10
- [Setting OS X deployment target](https://developer.apple.com/library/content/documentation/DeveloperTools/Conceptual/cross_development/Configuring/configuring.html)

File: cfep-02.md
Comment (jakirkham) [inline cfep-02.md]: We never have had support for these as we were always using the [`beta-xcode6.1`](https://docs.travis-ci.com/user/osx-ci-environment/#Xcode-6.1) image, which only has the 10.9 and 10.10 SDKs.

**Comment by inducer** [comment]
jakirkham notifications@github.com writes:

> Other relevant references.

Added, thx!

File: cfep-02.md
Comment (jakirkham) [inline cfep-02.md]: Also in need of this image are
- [Apache Arrow](https://github.com/conda-forge/conda-forge.github.io/issues/249#issuecomment-254331475)
- [`libdynd`](https://github.com/conda-forge/staged-recipes/pull/1051#issuecomment-233279062)

File: cfep-02.md
Comment (jakirkham) [inline cfep-02.md:20]: Would link this PR ( https://github.com/conda-forge/staged-recipes/pull/1094 ).

File: cfep-02.md
Comment (jakirkham) [inline cfep-02.md:18]: Simply need to change this [line](https://github.com/conda-forge/conda-smithy/blob/b1d1730c4b8c8dc849464b1b39569b0e61ec3130/conda_smithy/templates/travis.yml.tmpl#L12). So also easy. 😄

File: cfep-02.md
Comment (jakirkham) [inline cfep-02.md]: Should be `10.10` as it doesn't have the `10.11` SDK FWICT.

File: cfep-02.md
Comment (jakirkham) [inline cfep-02.md:39]: We have until `January 20th, 2017`. Would actually include that date verbatim as I don't think many people know this.

**Comment by inducer** [comment]
@jakirkham Please feel free to push directly to the text. Added you to the repo.

File: cfep-02.md
Comment (jakirkham) [inline cfep-02.md:24]: Don't think this is required. However, the most recent document I can find on C++ ABI compatibility is discussing 10.3.8 and before vs. 10.3.9 and after. Those OSes (and their associated compilers) are so ancient this seems irrelevant, but have included the link anyways.

xref: https://developer.apple.com/library/content/documentation/DeveloperTools/Conceptual/CppRuntimeEnv/Articles/CPPROverview.html

**Comment by jakirkham** [comment]
Thanks for writing all of this up, @inducer ! This is very helpful. Hopefully this will help get this important discussion moving.

**Comment by jakirkham** [comment]
Would also appreciate your feedback on this.

cc @izaid @insertinterestingnamehere @wesm

File: cfep-02.md
Comment (jakirkham) [reply cfep-02.md:39]: Noted below.

**Comment by inducer** [comment]
@jakirkham Thanks for your comments. I think I've got most of the massaged into the text.

**Comment by wesm** [comment]
+1

File: cfep-02.md
Comment (jakirkham) [inline cfep-02.md:71]: Probably should add this recent one too.

xref: https://blog.travis-ci.com/2016-11-17-retiring-some-osx-images/

**Comment by inducer** [comment]
jakirkham notifications@github.com writes:

> Probably should add this recent one too.
> 
> xref: https://blog.travis-ci.com/2016-11-17-retiring-some-osx-images/

Done.

**Comment by jakirkham** [comment]
Also brought this up at an informal meeting last week. It should come up in the formal meeting this week. I'd recommend anyone interested in this issue to go if they can. Unfortunately it falls on Thanksgiving US. So if people are unable to make it, I completely understand. Sadly I know I won't be able to make it either. Details linked below.

xref: https://conda-forge.hackpad.com/conda-forge-meetings-2YkV96cvxPG#:h=2016-11-24:-General-Discussion

**Comment by minrk** [comment]
I've started seeing this message on Travis:

screen shot 2016-11-25 at 12 47 53

which suggests that time is of the essence, and Travis' previous messaging about 6.1 being safe until January might no longer be accurate.

I also opened https://github.com/conda-forge/conda-smithy/issues/391, sibling to https://github.com/conda-forge/staged-recipes/pull/1094, which can be merged once this is agreed upon.

And based on the info linked here, I wrote [this script](https://gist.github.com/f13f3c7f8668134db9796daa643a57b7) for inspecting the deployment target of dylibs, inspired a bit by `conda inspect linkages`. We can try adding it somewhere so that it can be called during the test phase on Travis, but I'm not 100% sure where the right place would be to open a PR (smithy?, toolchain? conda-build?).

File: cfep-02.md
Comment (minrk) [inline cfep-02.md:61]: "by hand" is perhaps an overstatement - the toolchain package (mentioned above) already does this, and staged-recipes already includes a note in the build section that any package that compiles things should pull in toolchain as a build dependency.

**Comment by insertinterestingnamehere** [comment]
OSX development isn't my primary area of expertise, but FWIW I think this is a great idea.

**Comment by jakirkham** [comment]
Also have been seeing that @minrk . It is unclear to me whether this is a warning that they added to all the images being deprecated (of which there are several) or if it has meaning in this case. I think we still get the image until January. However, the date it drops has become a bit fuzzy.

xref: https://github.com/travis-ci/travis-ci/issues/6765#issuecomment-261826869

**Comment by jakirkham** [comment]
To help ease some concerns, the Travis CI `beta-xcode6.1` image will remain until January 20th.

xref: https://github.com/travis-ci/travis-ci/issues/6765#issuecomment-263294723

cc @minrk

**Comment by minrk** [comment]
@jakirkham thanks for checking!

I think this should be a pretty safe upgrade to make. Any recipe that pulls in toolchain should already be set.  The main thing to do is to ensure that, as packages are updated, make sure toolchain is pulled in as a dependency for those that don't have this already. We can do this at the human level (easiest, highest chance of occasional broken package on 10.9-only), or wait until a verification script like the one I linked above can be run automatically on build on OS X. I'm happy to start a PR for that, but I'm not sure yet which is the right repo to do it.

**Comment by inducer** [comment]
What's the criterion for officially calling this 'accepted'? Is there an official voting process? IMO, this is mostly a no-brainer, since the choice is mainly between "Do we wait a month until Travis forces our hand, or do we do this step-by-step on our own terms?'

File: cfep-02.md
Comment (pelson) [inline cfep-02.md]: @ocefpaf - I'd appreciate your input on this point.

File: cfep-02.md
Comment (pelson) [inline cfep-02.md:48]: Could you add a one-sentence context that summarises the need?

**Review by pelson — COMMENTED** [review]
This looks great to me @inducer. Thanks for taking the time to write this down. We still lack governance on "[consensus](https://github.com/conda-forge/conda-forge-enhancement-proposals/blob/master/cfep-01.md#resolution)", but given there is a cliff edge approaching, we need to time-bound the process so that we have a decision in the next few weeks (i.e. before 2017) so that we can start to implement/percolate the change across conda-forge.

The huge unknown is still

```

* Rebuild all packages with this? (TBD? Opinions?)

Likely not necessary given C++ ABI Compatibility.

```

If we can resolve this one piece of information definitively, we will know roughly how much work this whole proposal will entail (and indeed, how much work is lurking for us if we *do* drop off the cliff on the 20th Jan).

File: cfep-02.md
Comment (inducer) [reply cfep-02.md:48]: Done.

File: cfep-02.md
Comment (ocefpaf) [reply cfep-02.md]: I am against the use of any un-versioned package that can change the build characteristics like that. We already have have feedstocks that have to override those variables to build properly. IMO those belong in the recipe and are, ultimately, the maintainers decision.

With that said I no longer have a stake on this as we diverged from conda-forge a long time ago, since the toolchain was first added. Unfortunately most of earth-science software we build locally must use `gcc` and must provide `libgomp` and others at run time to ensure they will run anywhere.

Comment (inducer) [reply cfep-02.md]: So what do you suggest? Would you prefer a versioned dependency on `toolchain`? (And isn't each maintainer still at liberty to decide whether they want a versioned or unversioned dependency?) Or should each maintainer figure out setting things like the build target by hand? (And who catches it if someone forgets?)

Comment (ocefpaf) [reply cfep-02.md]: `> So what do you suggest?

Recommend but do not enforce. Many add the toolchain without even knowing what is the impact of those flags to the builds.

> And who catches it if someone forgets?

Correct me if I am wrong but the build target is not too bad if an old one is used, right?

Anyways, like I said above, I no longer have a stake on this and I don't really care what is decided here. I am following @pelson's own workflow to re-build everything for my day job. Also, many of the users that had issues with the toolchain, that broke the GIS stack, already left `conda-forge`.

**Comment by wesm** [comment]
What's our timeline to resolve this? January 21 is only a few weeks away

**Comment by jakirkham** [comment]
I put it on the meeting agenda.

File: cfep-02.md
Comment (jakirkham) [reply cfep-02.md]: I still advocate the `toolchain` and it is versioned. We can pin the versions in the feedstocks. We can even add the pinned version to the pinning script. That all being said, I feel like Filipe and I have had this exact conversation over and over without a different result. So am not eager to dive into it again.

Instead I would like to find a reasonable middle ground that everyone can agree on and is easy to implement. As we do actually need to set the Mac OS X deployment target. My proposal would be to set the `MACOSX_DEPLOYMENT_TARGET` in `conda-forge-build-setup` as show in PR ( https://github.com/conda-forge/conda-forge-build-setup-feedstock/pull/46 ). @msarahan helped me with a PR to whitelist the `MACOSX_DEPLOYMENT_TARGET` in `conda-build`. So this will work with `conda-build` 2.0.12+.

I think this proposal is nice for a few reasons. First it sets the deployment target everywhere immediately. So we don't have to worry about people who are unaware of this change accidentally breaking compatibility. Second it consolidates the effort between upgrading to `conda-build` 2 and upgrading the Travis CI image. We need `conda-build` 2 to leverage this feature. So the two efforts should not be "competing" for time, but working together. Third it moves the discussion away from compiler preferences to maintaining compatibility of binaries. As compatibility is the real concern we are dealing with here, it makes sense to make that the focus and this seems to the easiest most straightforward way to do that.

Comment (inducer) [reply cfep-02.md]: :+1: I've removed the `toolchain` wording and replaced it by a summary of what you've suggested.

File: cfep-02.md
Comment (jakirkham) [reply cfep-02.md:61]: After some discussion at our meeting yesterday, we have added PR ( https://github.com/conda-forge/conda-forge-build-setup-feedstock/pull/46 ), which forces the minimum build target on OS X to 10.9. While personally I still recommend the `toolchain` package and it does have broad usage throughout `conda-forge`, having the deployment target set in `conda-forge-build-setup` guarantees that everything builds in a compatible way. This should help further reduce the burden on feedstock maintainers by setting compatibility at the beginning of each build.

File: cfep-02.md
Comment (jakirkham) [reply cfep-02.md]: Thanks @inducer .

File: cfep-02.md
Comment (jakirkham) [inline cfep-02.md]: Do we need to change the status?

**Review by jakirkham — APPROVED** [review]
LGTM

Not sure if we need to change the status to "Accepted", but this looks like a complete proposal ready to merge.

**Comment by jakirkham** [comment]
Thanks for your work on this, @inducer .

**Comment by wesm** [comment]
Thanks everyone.

**Comment by inducer** [comment]
Updated the header to reflect acceptance.

**Comment by jakirkham** [comment]
Any final thoughts, @conda-forge/core ?

**Comment by mwcraig** [comment]
Looks great!

**Comment by jakirkham** [comment]
Alright, given as we have 10 days to get this sorted and I'm not hearing any complaints about this revision, am going to go ahead and merge it. Though if there are corrections, please feel free to raise issues and PRs.

**Comment by jakirkham** [comment]
Thanks @inducer for putting this together and thanks everyone for the thorough review. 👍

**Pull request merged**
